book/course/containers.md

```diff
@@ -0,0 +1,330 @@
+That makes them perfect for HPC, where we want all your work to run as your
+user.
+
+## Singularity vs Apptainer
+
+First there was Singularity, then there was commercialisation that kept an open
+source version, then a while after a fork after a slight disagreement, and
+Apptainer was born.  Apptainer has been developed from Singularity, and we'll
+only be looking at Apptainer in this course, but much of what's listed will
+equally apply to Singularity.  Whilst there remains a community edition of
+Singularity, we'll only be looking at Apptainer, but have tended to use the
+`singularity` command line, which is still provided for backwards
+compatibility.  Anywhere we use the `singularity` command, you could instead
+use `apptainer`.
+
+## Example of running a container
+
+It's always good to do something simple to start iwht, to prove to yourself
+things are working:
+
+```bash
+# Load the legacy singularity module
+$ module add singularity
+
+# Download a simple container
+$ singularity pull docker://godlovedc/lolcow
+
+# Run it
+$ singularity run lolcow_latest.sif
+INFO:    Using cached SIF image
+ _________________________________________
+/ This night methinks is but the daylight \
+| sick.                                   |
+|                                         |
+| -- William Shakespeare, "The Merchant   |
+\ of Venice"                              /
+ -----------------------------------------
+        \   ^__^
+         \  (oo)\_______
+            (__)\       )\/\
+                ||----w |
+                ||     ||
+```
+
+It might not look on the face of it that this is very impressive, but we've just:
+
+- pulled down a container from docker hub
+- converted it into a SIF image
+- run a piece of software, within an Ubuntu container
+
+### What do those steps mean?
+
+#### Pulling a container image down from Docker Hub
+
+Docker Hub is a repository of containers.  Pulling down a container involves
+pulling down the pieces of this container image over to your machine.
+
+#### Converting it into a SIF image
+
+Docker uses a container format that involves a number of layers, that add up
+together to form your final container image.  Singularity uses a different
+format SIF, which is a single file that contains everything you need to run
+your container.
+
+#### Running software within a container
+
+When you come to use a container, you can run software inside the container,
+but also see file shares outside the container, and access resources like
+network and GPU of the host machine.  This allows us to use all the power of an
+HPC, but with a software environment we've brought with us from elsewhere.  In
+the example we just used `run` which executes a predefined piece of software
+within the container, but you are not limited to only running the command set
+when the container was created.
+
+Note that you normally have read only access to the container, and only have
```

Comment:
Worth making this a note admonition?
````
```{note}
You normally have read only access to the container, and only have
write access to directories of the host mapped into the container.
```
````

Reply:
Fixed in 3168c5325f209687146caa461a39119794400b15